src/enchant-lsmod.1
.TH ENCHANT-LSMOD 1 "August 2017"
.SH NAME
Enchant \- list provider and dictionary information
.SH SYNOPSIS
.ll +8
.B enchant-lsmod
[[\fB\-lang\fR|\fB-word-chars\fR] [\fBlanguage_tag\fR]|\fB\-list-dicts\fR|\fB\-help\fR|\fB\-version\fR]
.ll -8
.br
.SH DESCRIPTION
.B enchant-lsmod
gives information about Enchant's spell-checker providers and the dictionaries each offers.
See \fBenchant\fR(1) for details about how the providers and dictionaries are located.
If no command-line options are given, the available providers are listed.
.SS OPTIONS
.TP
.B "\-lang"
Show which provider and dictionary will be used for the given language, or all languages if none is supplied.
.TP
.B "\-word\-chars"
Show the extra word characters for the given language, if available. This is of little interest to most users.
.TP
.B "\-list\-dicts"
List the provider and dictionary for all available languages.
.TP
.B "\-help"
Show brief help.
.TP
.B "\-version"
Prints the program's version.
.SH "SEE ALSO"
.BR enchant(1)
.SH "AUTHOR"
Written by Dom Lachowicz and Reuben Thomas.


NEWS
2.1.0 (August 7, 2017)
----------------------

This is a minor feature release. It adds the -word-chars flag to
enchant-lsmod, for a convenient way for other programs to get the extra word
characters for a particular dictionary.

A man page for enchant-lsmod is also now provided.

A little more code cleanup has been done, in particular by upstreaming
gnulib patches.


2.0.0 (August 4, 2017)
----------------------

The major version number has been incremented owing to API/ABI changes, but
in practice upgrading from 1.6.x should be easy.

Previously-deprecated APIs have been removed.

The little-used enchant_broker_get/set_param calls have been removed.

Some trivial API changes have been made to fix otherwise-unavoidable
compilation warnings both in libenchant and in application code. This is
strictly an ABI change (although the ABI may not actually have changed,
depending on the platform).

The provider API has been changed slightly: enchant_get_user_language is now
a documented public API (before it was marked private, but it has in fact
been exported for some years). enchant_get_user_config_dirs is now
enchant_get_user_config_dir, and returns only a single directory.

The plethora of configuration options previously available has been
rationalised and documented. In particular, support for relocation (so that
Enchant, or an application of which it is part, can be installed anywhere in
a filing system) has been rewritten and documented (see INSTALL).

The Myspell backend has been renamed to Hunspell to match the upstream
project. Users with their own enchant.ordering files will need to change
“myspell” to “hunspell”.


1.6.1 (February 6, 2017)
------------------------

Improvements to the enchant-ispell front-end, which is now a working ispell replacement.
Unit tests run on all platforms.
Various bug fixes and code clean-up.


1.6.0 (April 1, 2010)
---------------------

Fix bug 12567: the ispell sources aren't licensed under the LGPL
Add a function to get enchant's version (enchant_get_version)
Disable zemberek plugin by default, as it's known to cause issues/crashes with WebKit
Fix bug 12472: Win32 DLL dependency not found popup occurs when module has unmet dependencies
Possibly fix Ubuntu bug 474062
Fix bug 12409: Registry handle not closed in enchant_get_registry_value_ex
Fix bug 12406: Leak in _enchant_get_user_home_dirs() on Windows
Fix bug 12007: Update FSF address
Fix bug 12305: Zemberek module lists a Turkish dictionary even without Zemberek installed
Don't assert if passed a null string list
Fix bug 12350: enchant_pwl_init_with_file truncates pwl file
Fix a double-free memory corruption bug
Fix bug 12173: fix some small memory leaks
Fix bug 12174: mis-acceptence of dictionaries which start with a partial match of the lang id
Fix bug 12160: enchant 1.5.0 always looks in "lib" dir for plugins
Fix the build with the MSVC compiler
Add a --with-system-myspell option
Package missing compile-resource file
Compare paths ignoring case sensitivity on windows


1.5.0 (May 23, 2009)
--------------------

(News missing!)


1.4.2 (May 5, 2008)
-------------------

Voikko (Finnish) language support
Zemberek (Turkish) language support
Better support for Unicode in the personal dictionaries
Personal dictionaries offer better suggestions
Use OpenOffice's dictionaries on Windows
Aspell works on Windows
Can use a system-wide Hunspell/Myspell installation on Unix-like platforms
Require Hunspell 1.2.1
.NET bindings
More lax language matching rules (eg. if you request a "pl" dictionary, but
only have a "pl_PL" myspell dictionary installed, it will do the right
thing)
Use XDG's data-dirs spec for locating dictionaries (eg. ~/.config/enchant/myspell/)
Lots of unit tests
Lots of bug fixes


0.4.0
-------------------
Ispell-like command line interface


0.3.0
-------------------
Better Aspell detection
Aspell backend sync's dictionary lists to disk
Ispell crash fixes
Myspell crash fixes
Refined C++ API/ABI
Refined C ABI
Implemented some new APIs
Better implemented some internals


0.2.0
-------------------
Myspell support
Better Uspell support
Ability to enumerate providers
Win32 support


0.1.0
-------------------
Initial Release
Aspell/Pspell support
Ispell support
Uspell support
Ability to order dictionaries/providers




src/enchant-lsmod.c
/* enchant
 * Copyright (C) 2003 Dom Lachowicz
 * Copyright (C) 2017 Reuben Thomas <rrt@sc3d.org>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02110-1301, USA.
 *
 * In addition, as a special exception, Dom Lachowicz
 * gives permission to link the code of this program with
 * the non-LGPL Spelling Provider libraries (eg: a MSFT Office
 * spell checker backend) and distribute linked combinations including
 * the two.  You must obey the GNU Lesser General Public License in all
 * respects for all of the code used other than said providers. If you modify
 * this file, you may extend this exception to your version of the
 * file, but you are not obligated to do so.  If you do not wish to
 * do so, delete this exception statement from your version.
 */

#include "config.h"

#include <glib.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "unused-parameter.h"

#include "enchant.h"
#include "enchant-provider.h"

static void
describe_dict (const char * const lang_tag,
	       const char * const provider_name,
	       const char * const provider_desc _GL_UNUSED_PARAMETER,
	       const char * const provider_file _GL_UNUSED_PARAMETER,
	       void * user_data _GL_UNUSED_PARAMETER)
{
	printf ("%s (%s)\n", lang_tag, provider_name);
}

static void
describe_word_chars (const char * const lang_tag _GL_UNUSED_PARAMETER,
		     const char * const provider_name _GL_UNUSED_PARAMETER,
		     const char * const provider_desc _GL_UNUSED_PARAMETER,
		     const char * const provider_file _GL_UNUSED_PARAMETER,
		     void * user_data)
{
	EnchantDict *dict = (EnchantDict *)user_data;
	const char *word_chars = "";
	if (dict)
		word_chars = enchant_dict_get_extra_word_characters(dict);
	printf ("%s\n", word_chars ? word_chars : "");
}

static void
describe_provider (const char * name,
		   const char * desc,
		   const char * file _GL_UNUSED_PARAMETER,
		   void * user_data _GL_UNUSED_PARAMETER)
{
	printf ("%s (%s)\n", name, desc);
}

static void
usage (const char *progname)
{
	fprintf (stderr, "%s [[-lang|-word-chars] [language_tag]|-list-dicts|-help|-version]\n", progname);
}

int
main (int argc, char **argv)
{
	EnchantBroker *broker = enchant_broker_init ();
	char * lang_tag = NULL;
	int retcode = 0;

	if (argc > 1) {
		if (!strcmp (argv[1], "-lang") || !strcmp(argv[1], "-word-chars")) {
			if (argc > 2) {
				lang_tag = strdup (argv[2]);
			} else {
				lang_tag = enchant_get_user_language();

				if (!lang_tag || !strcmp (lang_tag, "C")) {
					free(lang_tag);
					lang_tag = strdup ("en");
				}
			}
			if (!lang_tag) {
				fprintf (stderr, "Error: language tag not specified and environment variable $LANG not set\n");
				retcode = 1;
			} else {
				EnchantDict *dict = enchant_broker_request_dict (broker, lang_tag);
				if (!dict) {
					fprintf (stderr, "No dictionary available for '%s'\n", lang_tag);
					retcode = 1;
				} else {
					enchant_dict_describe (dict,
							       !strcmp (argv[1], "-lang") ? describe_dict : describe_word_chars,
							       dict);
					enchant_broker_free_dict (broker, dict);
				}
			}
		} else if (!strcmp (argv[1], "-h") || !strcmp(argv[1], "-help")) {
			usage (argv[0]);
		} else if (!strcmp (argv[1], "-v") || !strcmp (argv[1], "-version")) {
			fprintf (stderr, "%s %s\n", argv[0], PACKAGE_VERSION);
		} else if (!strcmp (argv[1], "-list-dicts")) {
			enchant_broker_list_dicts (broker, describe_dict, NULL);
		} else {
			fprintf (stderr, "Invalid argument %s\n", argv[1]);
			usage (argv[0]);
			retcode = 1;
		}
	} else
		enchant_broker_describe (broker, describe_provider, NULL);

	free (lang_tag);
	enchant_broker_free (broker);
	return retcode;
}
